### PR TITLE
Fix/subapps blocking middleware

### LINKED ISSUE #433: app.route blocks middleware mounted at a subpath
# Case 1
## To Reproduce
```ts
import { App } from '@tinyhttp/app'

const app = new App()

app.route('/')
  .get((req, res) => void res.send(`this is rooty`))

app.use('/foo', (req, res) => void res.send(`this is not rooty`))

app.listen(3000)
```

## Result
- [http://localhost:3000](http://localhost:3000) yields `this is rooty`
- [http://localhost:3000/foo](http://localhost:3000/foo) yields `Not Found`

## Expected behavior
- [http://localhost:3000](http://localhost:3000) yields `this is rooty`
- [http://localhost:3000/foo](http://localhost:3000/foo) yields `this is not rooty`

# Case 2 
## To Reproduce
```ts
import { App } from '@tinyhttp/app'

const app = new App()

app.use('/', (req, res) => void res.send(`this is rooty`))

app.use('/foo', (req, res) => void res.send(`this is not rooty`))

app.listen(3000)
```

## Result
- [http://localhost:3000](http://localhost:3000) yields `this is rooty`
- [http://localhost:3000/foo](http://localhost:3000/foo) yields `this is rooty`

## Expected behavior
- [http://localhost:3000](http://localhost:3000) yields `this is rooty`
- [http://localhost:3000/foo](http://localhost:3000/foo) yields `this is not rooty`

## Evironment

- `node`: 20.11.1
- `@tinyhttp/app`: 2.3.0

### PR DESCRIPTION
**Pull request by Lordfirespeed**
Resolves #433

### CODE REVIEW

File: tests/core/app.test.ts [hunk 1]
```diff
@@ -683,6 +685,23 @@ describe('Subapps', () => {
 
     await fetch('/').expect(200, 'Hello World!')
   })
+  it('sub-app mounted on root and subpath', async () => {
```
Comment (aarontravass) [inline tests/core/app.test.ts:688]: Can you rename this test to reflect the actual test?

File: packages/app/src/app.ts [hunk 2]
```diff
@@ -387,17 +391,30 @@ export class App<Req extends Request = Request, Res extends Response = Response>
 
     let idx = 0
 
-    const loop = () => res.writableEnded || (idx < mw.length && handle(mw[idx++])(req, res, next))
+    const handleNext = (): void => void handle(mw[idx++])(req, res, next)
```
Comment (aarontravass) [inline packages/app/src/app.ts:394]: Type annotation can be removed.

**Review by aarontravass — CHANGES_REQUESTED** [review]

**Comment by aarontravass** [comment]
Can you add more verb specific (`app.route`, `app.all` etc) test cases?

File: tests/core/app.test.ts (hunk 1 above)
Comment (Lordfirespeed) [reply tests/core/app.test.ts:688]: I don't like that name either - just trying my best to follow the convention that the similar tests follow.

If you give me the particular name you want me to use, sure. 

I don't fancy playing that guessing game presently 😅

File: packages/app/src/app.ts (hunk 2 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:394]: it *can*, but why be implicit when you can be explicit? 

Annotating `void` ensures that the signature is not accidentally changed in future.

**Comment by Lordfirespeed** [comment]
> Can you add verb specific (`app.route`, `app.all` etc) test cases? 

`.route` is an alias for `.use(new App)`, so I don't see the point.

`.all`, `.get` etc are out of scope for this PR, as explained in #433

**Comment by aarontravass** [comment]
> Yes, I'll add a test for one verb-specific method (`get`) and the verb catch-all method

Sounds good. Thanks.

File: tests/core/app.test.ts (hunk 1 above)
Comment (aarontravass) [reply tests/core/app.test.ts:688]: Maybe something along the lines of `should search for routes in other subapps even if a subapp mounted on '/' is present`. What do you reckon?

**Comment by aarontravass** [comment]
> `.route` is an alias for `.use(new App)`, so I don't see the point.

To be honest, I wouldn't mind adding it as a sanity check.

File: tests/core/app.test.ts (hunk 1 above)
Comment (Lordfirespeed) [reply tests/core/app.test.ts:688]: Aye, I catch your drift. I might go for something like this:
```ts
describe("when a subapp is mounted on '/'", () => {
  it("should continue middleware execution when subapp middleware is exhausted", () => { ... })
  // other related tests
})
```

Comment (aarontravass) [reply tests/core/app.test.ts:688]: Great! Are there other test cases which you can group under `when a subapp is mounted on '/'`? It wouldn't make sense to have a test suite with only 1 test.

Comment (Lordfirespeed) [reply tests/core/app.test.ts]: > Great! Are there other test cases which you can group under `when a subapp is mounted on '/'`? It wouldn't make sense to have a test suite with only 1 test.

There are, so this is moot, but regardless - I disagree, using `describe` to provide context to a single testcase is a common practise. See Mocha's `context` ([example](https://mochajs.org/#bdd>))

**Comment by Lordfirespeed** [comment]
My only remaining concern at this point is that it's possible I've significantly (detrimentally) affected framework benchmarks - I'm not sure how to test whether or not that is the case, though 😅

File: tests/core/app.test.ts (hunk 1 above)
Comment (aarontravass) [reply tests/core/app.test.ts]: Cool.

**Comment by talentlessguy** [comment]
@Lordfirespeed  you can use https://github.com/mcollina/autocannon but keep in mind that it's results are unreliable because network has a bunch of side effects

File: packages/app/src/app.ts [hunk 3]
```diff
@@ -79,7 +79,8 @@ export class App<Req extends Request = Request, Res extends Response = Response>
       ...options.settings
     }
     this.applyExtensions = options?.applyExtensions
-    this.attach = (req, res) => setImmediate(this.handler.bind(this, req, res, undefined), req, res)
+    const boundHandler = this.handler.bind(this)
+    this.attach = (req, res, next?: NextFunction) => setImmediate(boundHandler, req, res, next)
```
Comment (talentlessguy) [inline packages/app/src/app.ts:83]: no need to move it to a separate variable cuz it's used only once

File: packages/app/src/app.ts [hunk 4]
```diff
@@ -152,7 +153,7 @@ export class App<Req extends Request = Request, Res extends Response = Response>
 
   /**
    * Render a template
-   * @param file What to render
+   * @param name What to render
```
Comment (talentlessguy) [inline packages/app/src/app.ts:156]: seems like unrelated change

File: packages/app/src/app.ts [hunk 5]
```diff
@@ -387,17 +391,30 @@ export class App<Req extends Request = Request, Res extends Response = Response>
 
     let idx = 0
 
-    const loop = () => res.writableEnded || (idx < mw.length && handle(mw[idx++])(req, res, next))
```
Comment (talentlessguy) [inline packages/app/src/app.ts:390]: a function rename is unnecessary

File: packages/app/src/app.ts [hunk 6]
```diff
@@ -387,17 +391,30 @@ export class App<Req extends Request = Request, Res extends Response = Response>
...
+    }
 
-    loop()
+    handleNext()
   }
 
   /**
    * Creates HTTP server and dispatches middleware
    * @param port server listening port
-   * @param Server callback after server starts listening
```
Comment (talentlessguy) [inline packages/app/src/app.ts:400]: unrelated change

**Review by talentlessguy — CHANGES_REQUESTED** [review]

File: packages/app/src/app.ts (hunk 3 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:83]: The difference is that `.bind` is invoked on every invocation when the variable is inlined - in theory, what I have done saves memory

File: packages/app/src/app.ts (hunk 6 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:400]: My IDE was giving me grief about the jsdoc parameter being missing. Moving this change (and the other jsdoc one above) to another PR would be a waste of my time.

File: packages/app/src/app.ts (hunk 4 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:156]: My IDE was giving me grief about the jsdoc parameter being missing.

File: packages/app/src/app.ts (hunk 5 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:390]: Happy to revert, but maybe we can agree on a more suitable name? Both 'loop' and 'handleNext' are not great at explaining what it does.

Comment (talentlessguy) [reply packages/app/src/app.ts:390]: I don't know about a more suitable name but renaming stuff isn't in the scope of this PR. The PR is fixing a bug.

File: packages/app/src/app.ts (hunk 6 above)
Comment (talentlessguy) [reply packages/app/src/app.ts:400]: ok

File: packages/app/src/app.ts (hunk 5 above)
Comment (Lordfirespeed) [reply packages/app/src/app.ts:390]: That logic is really flawed - in order to fix a bug, often you have to refactor things. Trying to keep names the same whilst changing behaviour means intent/behaviour will diverge from the variable/function names, reducing code quality.

You've expressed you aren't interested in trying to agree on a better name, so I'll revert to `loop`

Comment (talentlessguy) [reply packages/app/src/app.ts:390]: it's not flawed, the function name "loop" corresponds to middleware dispatch loop. That's why it's called as such. It was taken from Polka's source where it does the same thing and is called the same

Comment (Lordfirespeed) [reply packages/app/src/app.ts:390]: No, that's not what I mean. I see why it's called `loop` and I think that logic is sound (though I don't think the name is perfect).

The logic of 'this PR is a fix, so we can't rename anything' is what I am referring to.

**Comment by Lordfirespeed** [comment]
Gave `autocannon` a go and I see what you mean - it's quite variable - that said, I can't see significant differences between `master` and this PR, so performance concerns can probably be put aside.

**Review by talentlessguy — APPROVED** [review]
